### PR TITLE
[AC-2284] Set organization billing email to MSP billing email when linked

### PR DESCRIPTION
**Pull request by amorask-bitwarden**
## Type of change

```
- [X] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

This PR adds business logic that has been missing from the "Add client organization" process. Whenever an organization is linked to an MSP, that organization's billing email and Stripe account email needs to be set to the MSP's billing email.

https://github.com/bitwarden/server/assets/144709477/7141f202-4d09-41fd-8b39-b9673606f71a

## Code changes

* **ProviderService.cs:** Set the organization's billing email and Stripe account email to the MSP's billing email when the organization is added to the MSP.
* **ProviderServiceTests.cs:** Update the success case to account for the new invocations.

### CODE REVIEW

File: bitwarden_license/test/Commercial.Core.Test/AdminConsole/Services/ProviderServiceTests.cs [hunk 1]
```diff
@@ -458,17 +458,88 @@ await sutProvider.GetDependency<IEventService>().Received()
...
+
+        await organizationRepository.Received(1)
+            .ReplaceAsync(Arg.Is<Organization>(org => org.BillingEmail == provider.BillingEmail));
+
+        await sutProvider.GetDependency<IStripeAdapter>().Received(1).CustomerUpdateAsync(
+            organization.GatewayCustomerId,
+            Arg.Is<CustomerUpdateOptions>(options => options.Email == provider.BillingEmail));
+
+        await sutProvider.GetDependency<IEventService>()
+            .Received().LogProviderOrganizationEventAsync(Arg.Any<ProviderOrganization>(),
```
Comment (r-tome) [inline bitwarden_license/test/Commercial.Core.Test/AdminConsole/Services/ProviderServiceTests.cs:482]:
```suggestion
.Received(1).LogProviderOrganizationEventAsync(Arg.Is<ProviderOrganization>(po => po.ProviderId == provider.Id && po.OrganizationId == organization.Id && po.Key == key),
```

File: bitwarden_license/test/Commercial.Core.Test/AdminConsole/Services/ProviderServiceTests.cs [hunk 2]
```diff
@@ -458,17 +458,88 @@ await sutProvider.GetDependency<IEventService>().Received()
...
+
+        var providerOrganizationRepository = sutProvider.GetDependency<IProviderOrganizationRepository>();
+        providerOrganizationRepository.GetByOrganizationId(organization.Id).ReturnsNull();
+
+        var organizationRepository = sutProvider.GetDependency<IOrganizationRepository>();
+        organizationRepository.GetByIdAsync(organization.Id).Returns(organization);
+
+        await sutProvider.Sut.AddOrganization(provider.Id, organization.Id, key);
+
+        await providerOrganizationRepository.ReceivedWithAnyArgs().CreateAsync(default);
```
Comment (r-tome) [inline bitwarden_license/test/Commercial.Core.Test/AdminConsole/Services/ProviderServiceTests.cs:472]:
```suggestion
        await providerOrganizationRepository.Received(1).CreateAsync(Arg.Is<ProviderOrganization>(po =>
            po.ProviderId == provider.Id && po.OrganizationId == organization.Id && po.Key == key));
```

File: bitwarden_license/test/Commercial.Core.Test/AdminConsole/Services/ProviderServiceTests.cs [hunk 3]
```diff
@@ -458,17 +458,88 @@ await sutProvider.GetDependency<IEventService>().Received()
...
+        var providerOrganizationRepository = sutProvider.GetDependency<IProviderOrganizationRepository>();
+        var expectedPlanType = PlanType.EnterpriseAnnually;
+        organization.PlanType = PlanType.EnterpriseAnnually;
+        sutProvider.GetDependency<IOrganizationRepository>().GetByIdAsync(organization.Id).Returns(organization);
+
+        await sutProvider.Sut.AddOrganization(provider.Id, organization.Id, key);
+
+        await providerOrganizationRepository.ReceivedWithAnyArgs().CreateAsync(default);
+        await sutProvider.GetDependency<IEventService>()
+            .Received().LogProviderOrganizationEventAsync(Arg.Any<ProviderOrganization>(),
```
Comment (r-tome) [inline bitwarden_license/test/Commercial.Core.Test/AdminConsole/Services/ProviderServiceTests.cs:503]:
```suggestion
            .Received(1).LogProviderOrganizationEventAsync(Arg.Is<ProviderOrganization>(po => po.ProviderId == provider.Id && po.OrganizationId == organization.Id && po.Key == key),
```

File: bitwarden_license/test/Commercial.Core.Test/AdminConsole/Services/ProviderServiceTests.cs [hunk 4]
```diff
@@ -458,17 +458,88 @@ await sutProvider.GetDependency<IEventService>().Received()
...
+        sutProvider.GetDependency<IProviderRepository>().GetByIdAsync(provider.Id).Returns(provider);
+
+        var providerOrganizationRepository = sutProvider.GetDependency<IProviderOrganizationRepository>();
+        var expectedPlanType = PlanType.EnterpriseAnnually;
+        organization.PlanType = PlanType.EnterpriseAnnually;
+        sutProvider.GetDependency<IOrganizationRepository>().GetByIdAsync(organization.Id).Returns(organization);
+
+        await sutProvider.Sut.AddOrganization(provider.Id, organization.Id, key);
+
+        await providerOrganizationRepository.ReceivedWithAnyArgs().CreateAsync(default);
```
Comment (r-tome) [inline bitwarden_license/test/Commercial.Core.Test/AdminConsole/Services/ProviderServiceTests.cs:501]:
```suggestion
        await providerOrganizationRepository.Received(1).CreateAsync(Arg.Is<ProviderOrganization>(po =>
            po.ProviderId == provider.Id && po.OrganizationId == organization.Id && po.Key == key));
```

**Review by r-tome — CHANGES_REQUESTED** [review]
Looks good to me but I'd prefer if we avoid using `Arg.Any()` on the unit tests.

File: bitwarden_license/test/Commercial.Core.Test/AdminConsole/Services/ProviderServiceTests.cs (hunk 1 above)
Comment (amorask-bitwarden) [reply bitwarden_license/test/Commercial.Core.Test/AdminConsole/Services/ProviderServiceTests.cs:482]: Hi @r-tome, that specific invocation was using Arg.Any before my change - you can see my updates Arg.Is predicates. Happy to update though.

**Comment by amorask-bitwarden** [comment]
> Looks good to me but I'd prefer if we avoid using `Arg.Any()` on the unit tests.

Hi @r-tome, those usages of `Arg.Any` were in place before my update. All of the assertions I added use `Arg.Is` predicates. However, if you'd like me to update the existing ones unrelated to my change, I can do so.

**Comment by amorask-bitwarden** [comment]
> Looks good to me but I'd prefer if we avoid using `Arg.Any()` on the unit tests.

All Arg.Any or `default` args for test assertions related to `AddOrganization` method have been removed as part of https://github.com/bitwarden/server/pull/3897/commits/3533f52ea1cf13dbc8070903c2fa2e45a9af9272.

**Review by r-tome — APPROVED** [review]
@amorask-bitwarden thats fair and I see you've updated the unit tests. Nice to improve them while we're at it 😁

**Pull request merged**